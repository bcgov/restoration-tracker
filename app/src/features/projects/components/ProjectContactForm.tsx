import Button from '@material-ui/core/Button';
import Box from '@material-ui/core/Box';
import { Theme } from '@material-ui/core/styles/createMuiTheme';
import makeStyles from '@material-ui/core/styles/makeStyles';
import Typography from '@material-ui/core/Typography';
import { FieldArray, useFormikContext } from 'formik';
import React, { useState } from 'react';
import yup from 'utils/YupSchema';
import ProjectContactItemForm, {
  IProjectContactItemForm,
  IProjectContactItemFormProps,
  ProjectContactItemInitialValues,
  ProjectContactItemYupSchema
} from './ProjectContactItemForm';
import { Icon } from '@mdi/react';
import { mdiPencilOutline, mdiPlus, mdiTrashCanOutline } from '@mdi/js';
import EditDialog from 'components/dialog/EditDialog';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import Paper from '@material-ui/core/Paper';
import Divider from '@material-ui/core/Divider';
import Toolbar from '@material-ui/core/Toolbar';
import IconButton from '@material-ui/core/IconButton';
import Grid from '@material-ui/core/Grid';

export interface IProjectContactForm {
  contact: {
    contacts: IProjectContactItemForm[];
  };
}

export const ProjectContactInitialValues: IProjectContactForm = {
  contact: {
    contacts: []
  }
};

export const ProjectContactYupSchema = yup.object().shape({
  contact: yup.object().shape({
    contacts: yup.array().of(ProjectContactItemYupSchema)
  })
});

export type IProjectContactFormProps = IProjectContactItemFormProps;

const useStyles = makeStyles((theme: Theme) => ({
  legend: {
    marginTop: '1rem',
    float: 'left',
    marginBottom: '0.75rem',
    letterSpacing: '-0.01rem'
  },
  title: {
    flexGrow: 1,
    marginRight: '1rem',
    whiteSpace: 'nowrap',
    overflow: 'hidden',
    textOverflow: 'ellipsis',
    fontWeight: 700
  },
  titleDesc: {
    marginLeft: theme.spacing(1),
    fontWeight: 400
  },
  contactListItem: {
    padding: 0,
    '& + li': {
      marginTop: theme.spacing(2)
    }
  },
  contactListItemInner: {
    flexGrow: 1,
    flexShrink: 1,
    overflow: 'hidden'
  },
  contactListItemToolbar: {
    paddingRight: theme.spacing(2)
  }
}));

/**
 * Create project - contact section
 *
 * @return {*}
 */
const ProjectContactForm: React.FC<IProjectContactFormProps> = ({ coordinator_agency }) => {
  const classes = useStyles();

  const { values } = useFormikContext<IProjectContactForm>();

  const [isModalOpen, setIsModalOpen] = useState(false);

  // Tracks information about the contact that is being added/edited
  const [currentProjectContact, setCurrentProjectContact] = useState({
    index: 0,
    values: ProjectContactItemInitialValues
  });

  return (
    <>
      <Box mb={3} maxWidth={'72ch'}>
        <Typography variant="body1" color="textSecondary">
          Specify all contacts for the project.
        </Typography>
      </Box>
      <Box>
        <FieldArray
          name="contact.contacts"
          render={(arrayHelpers) => (
            <Box mb={2}>
              <EditDialog
                dialogTitle={'Add Contact'}
                open={isModalOpen}
                component={{
                  element: <ProjectContactItemForm coordinator_agency={coordinator_agency} />,
                  initialValues: currentProjectContact.values,
                  validationSchema: ProjectContactItemYupSchema
                }}
                onCancel={() => setIsModalOpen(false)}
                onSave={(projectContactValues) => {
                  // If current user is primary, set all other to false
                  if(projectContactValues.is_primary) {
                    values.contact.contacts.forEach((item, index) => {
                      if(index !== currentProjectContact.index) {
                        item.is_primary = 'false';
                      }
                    })
                  }

                  projectContactValues.is_primary = String(projectContactValues.is_primary)

                  if (currentProjectContact.index < values.contact.contacts.length) {
                    // Update an existing item
                    arrayHelpers.replace(currentProjectContact.index, projectContactValues);
                  } else {
                    // Add a new item
                    arrayHelpers.push(projectContactValues);
                  }

                  // Close the modal
                  setIsModalOpen(false);
                }}
              />
              <List dense disablePadding>
                {!values.contact.contacts.length && (
                  <ListItem dense component={Paper}>
                    <Box display="flex" flexGrow={1} justifyContent="center" alignContent="middle" p={2}>
                      <Typography variant="subtitle2">No Contacts</Typography>
                    </Box>
                  </ListItem>
                )}
                {values.contact.contacts.map((contact, index) => (
                  <ListItem dense className={classes.contactListItem} key={index}>
                    <Paper className={classes.contactListItemInner}>
                      <Toolbar className={classes.contactListItemToolbar}>
                        <Typography className={classes.title}>
                          {`${contact.first_name} ${contact.last_name}`}
                          <sup>
                            <Typography variant="caption" color="textSecondary">
                              {JSON.parse(contact.is_primary) ? ' Primary': ''}
                            </Typography>
                          </sup>
                        </Typography>

                        <IconButton
                          color="primary"
                          data-testid={'edit-button-' + index}
                          title="Edit Contact"
                          aria-label="Edit Contact"
                          onClick={() => {
                            setCurrentProjectContact({
                              index: index,
                              values: values.contact.contacts[index]
                            });
                            setIsModalOpen(true);
                          }}>
                          <Icon path={mdiPencilOutline} size={1} />
                        </IconButton>
                        <IconButton
                          color="primary"
                          data-testid={'delete-button-' + index}
                          title="Remove Contact"
                          aria-label="Remove Contact"
                          onClick={() => arrayHelpers.remove(index)}>
                          <Icon path={mdiTrashCanOutline} size={1} />
                        </IconButton>
                      </Toolbar>
                      <Divider />
                      <Box py={2} px={3}>
                        <Grid container spacing={2}>
                          <Grid item xs={12} sm={6} md={4}>
                            <Typography variant="body2" color="textSecondary">
                              Agency
                            </Typography>
                            <Typography variant="body1">{contact.agency}</Typography>
                          </Grid>
                          <Grid item xs={12} sm={6} md={4}>
                            <Typography variant="body2" color="textSecondary">
                              Email
                            </Typography>
                            <Typography variant="body1">{contact.email_address}</Typography>
                          </Grid>
                        </Grid>
                      </Box>
                    </Paper>
                  </ListItem>
                ))}
              </List>
            </Box>
          )}
        />
      </Box>
      <Button
        variant="outlined"
        color="primary"
        title="Add Contact"
        aria-label="Add Contact"
        startIcon={<Icon path={mdiPlus} size={1}></Icon>}
        data-testid="add-contact-button"
        onClick={() => {
          setCurrentProjectContact({
            index: values.contact.contacts.length,
            values: ProjectContactItemInitialValues
          });
          setIsModalOpen(true);
        }}>
        Add Contact
      </Button>
    </>
  );
};

export default ProjectContactForm;
