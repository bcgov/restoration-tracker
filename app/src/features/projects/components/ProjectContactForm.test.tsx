import { render } from '@testing-library/react';
import ProjectContactForm, {
  ProjectContactInitialValues,
  ProjectContactYupSchema
} from 'features/projects/components/ProjectContactForm';
import { Formik } from 'formik';
import React from 'react';

const handleSaveAndNext = jest.fn();

const agencies = ['Agency 1', 'Agency 2', 'Agency 3'];

const projectContactFilledValues = {
  contact: {
    contacts: [
      {
        first_name: 'Nerea',
        last_name: 'Oneal',
        email_address: 'quxu@mailinator.com',
        agency: 'Agency 3',
        is_public: 'true',
        is_primary: 'true'
      },
      {
        first_name: 'John',
        last_name: 'Doe',
        email_address: 'jd@mailinator.com',
        agency: 'Agency 4',
        is_public: 'true',
        is_primary: 'true'
      }
    ]
  }
};

describe('Project Contact Form', () => {
  it('renders correctly the empty component correctly', () => {
    const { asFragment } = render(
      <Formik
        initialValues={ProjectContactInitialValues}
        validationSchema={ProjectContactYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async (values) => {
          handleSaveAndNext(values);
        }}>
        {() => <ProjectContactForm coordinator_agency={[]} />}
      </Formik>
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('renders correctly the filled component correctly', () => {
    const { asFragment } = render(
      <Formik
        initialValues={projectContactFilledValues}
        validationSchema={ProjectContactYupSchema}
        validateOnBlur={true}
        validateOnChange={false}
        onSubmit={async (values, helper) => {
          handleSaveAndNext(values);
        }}>
        {() => <ProjectContactForm coordinator_agency={agencies} />}
      </Formik>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
