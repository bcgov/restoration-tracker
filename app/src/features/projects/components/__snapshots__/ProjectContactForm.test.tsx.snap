// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Project Contact Form renders correctly the empty component correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-7"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Specify all contacts for the project.
    </p>
  </div>
  <div
    class="MuiBox-root MuiBox-root-8"
  >
    <div
      class="MuiBox-root MuiBox-root-9"
    >
      <ul
        class="MuiList-root MuiList-dense"
      >
        <div
          class="MuiPaper-root MuiListItem-root MuiListItem-dense MuiListItem-gutters MuiPaper-elevation1 MuiPaper-rounded"
        >
          <div
            class="MuiBox-root MuiBox-root-10"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2"
            >
              No Contacts
            </h6>
          </div>
        </div>
      </ul>
    </div>
  </div>
  <button
    aria-label="Add Contact"
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
    data-testid="add-contact-button"
    tabindex="0"
    title="Add Contact"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium"
      >
        <svg
          role="presentation"
          style="width: 1.5rem; height: 1.5rem;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
            style="fill: currentColor;"
          />
        </svg>
      </span>
      Add Contact
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;

exports[`Project Contact Form renders correctly the filled component correctly 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root MuiBox-root-17"
  >
    <p
      class="MuiTypography-root MuiTypography-body1 MuiTypography-colorTextSecondary"
    >
      Specify all contacts for the project.
    </p>
  </div>
  <div
    class="MuiBox-root MuiBox-root-18"
  >
    <div
      class="MuiBox-root MuiBox-root-19"
    >
      <ul
        class="MuiList-root MuiList-dense"
      >
        <li
          class="MuiListItem-root makeStyles-contactListItem-14 MuiListItem-dense MuiListItem-gutters"
        >
          <div
            class="MuiPaper-root makeStyles-contactListItemInner-15 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiToolbar-root MuiToolbar-regular makeStyles-contactListItemToolbar-16 MuiToolbar-gutters"
            >
              <p
                class="MuiTypography-root makeStyles-title-12 MuiTypography-body1"
              >
                Nerea Oneal
                <sup
                  class="MuiBox-root MuiBox-root-20"
                >
                  <span
                    class="MuiTypography-root MuiTypography-caption MuiTypography-colorTextSecondary"
                  >
                    Primary
                  </span>
                </sup>
              </p>
              <button
                aria-label="Edit Contact"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="edit-button-0"
                tabindex="0"
                title="Edit Contact"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    role="presentation"
                    style="width: 1.5rem; height: 1.5rem;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                      style="fill: currentColor;"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Remove Contact"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="delete-button-0"
                tabindex="0"
                title="Remove Contact"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    role="presentation"
                    style="width: 1.5rem; height: 1.5rem;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z"
                      style="fill: currentColor;"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <hr
              class="MuiDivider-root"
            />
            <div
              class="MuiBox-root MuiBox-root-21"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                  >
                    Agency
                  </p>
                  <p
                    class="MuiTypography-root MuiTypography-body1"
                  >
                    Agency 3
                  </p>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                  >
                    Email
                  </p>
                  <p
                    class="MuiTypography-root MuiTypography-body1"
                  >
                    quxu@mailinator.com
                  </p>
                </div>
              </div>
            </div>
          </div>
        </li>
        <li
          class="MuiListItem-root makeStyles-contactListItem-14 MuiListItem-dense MuiListItem-gutters"
        >
          <div
            class="MuiPaper-root makeStyles-contactListItemInner-15 MuiPaper-elevation1 MuiPaper-rounded"
          >
            <div
              class="MuiToolbar-root MuiToolbar-regular makeStyles-contactListItemToolbar-16 MuiToolbar-gutters"
            >
              <p
                class="MuiTypography-root makeStyles-title-12 MuiTypography-body1"
              >
                John Doe
                <sup
                  class="MuiBox-root MuiBox-root-22"
                >
                  <span
                    class="MuiTypography-root MuiTypography-caption MuiTypography-colorTextSecondary"
                  >
                    Primary
                  </span>
                </sup>
              </p>
              <button
                aria-label="Edit Contact"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="edit-button-1"
                tabindex="0"
                title="Edit Contact"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    role="presentation"
                    style="width: 1.5rem; height: 1.5rem;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M14.06,9L15,9.94L5.92,19H5V18.08L14.06,9M17.66,3C17.41,3 17.15,3.1 16.96,3.29L15.13,5.12L18.88,8.87L20.71,7.04C21.1,6.65 21.1,6 20.71,5.63L18.37,3.29C18.17,3.09 17.92,3 17.66,3M14.06,6.19L3,17.25V21H6.75L17.81,9.94L14.06,6.19Z"
                      style="fill: currentColor;"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
              <button
                aria-label="Remove Contact"
                class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorPrimary"
                data-testid="delete-button-1"
                tabindex="0"
                title="Remove Contact"
                type="button"
              >
                <span
                  class="MuiIconButton-label"
                >
                  <svg
                    role="presentation"
                    style="width: 1.5rem; height: 1.5rem;"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M9,3V4H4V6H5V19A2,2 0 0,0 7,21H17A2,2 0 0,0 19,19V6H20V4H15V3H9M7,6H17V19H7V6M9,8V17H11V8H9M13,8V17H15V8H13Z"
                      style="fill: currentColor;"
                    />
                  </svg>
                </span>
                <span
                  class="MuiTouchRipple-root"
                />
              </button>
            </div>
            <hr
              class="MuiDivider-root"
            />
            <div
              class="MuiBox-root MuiBox-root-23"
            >
              <div
                class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2"
              >
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                  >
                    Agency
                  </p>
                  <p
                    class="MuiTypography-root MuiTypography-body1"
                  >
                    Agency 4
                  </p>
                </div>
                <div
                  class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4"
                >
                  <p
                    class="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary"
                  >
                    Email
                  </p>
                  <p
                    class="MuiTypography-root MuiTypography-body1"
                  >
                    jd@mailinator.com
                  </p>
                </div>
              </div>
            </div>
          </div>
        </li>
      </ul>
    </div>
  </div>
  <button
    aria-label="Add Contact"
    class="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary"
    data-testid="add-contact-button"
    tabindex="0"
    title="Add Contact"
    type="button"
  >
    <span
      class="MuiButton-label"
    >
      <span
        class="MuiButton-startIcon MuiButton-iconSizeMedium"
      >
        <svg
          role="presentation"
          style="width: 1.5rem; height: 1.5rem;"
          viewBox="0 0 24 24"
        >
          <path
            d="M19,13H13V19H11V13H5V11H11V5H13V11H19V13Z"
            style="fill: currentColor;"
          />
        </svg>
      </span>
      Add Contact
    </span>
    <span
      class="MuiTouchRipple-root"
    />
  </button>
</DocumentFragment>
`;
